[LMS-323] Updates to Core Collection In Prep for Engagement Starter Kit

stephenfuqua opened the pull request:
The full plan is described in https://techdocs.ed-fi.org/display/EDFIODS/Data+Model+and+Metrics.

stephenfuqua commented on src/EdFi.AnalyticsMiddleTier.DataStandard2/Base/MSSQL/0030-View-SectionDim-for-Student-Engagement.sql:
```diff
@@ -0,0 +1,57 @@
+﻿-- SPDX-License-Identifier: Apache-2.0
```
![image](https://user-images.githubusercontent.com/9324390/125664224-1b440640-547d-48ef-a228-ed2673086a5a.png)

I was able to get rid of `DISTINCT`  by introducing the join condition for `CourseOffering.TermDescriptorId`.

stephenfuqua replied:
The change to `LocalEducationAgencyId` is now in Slack for discussion.

stephenfuqua replied:
`LocalEducationAgencyId` has now been restored.

stephenfuqua commented on src/EdFi.AnalyticsMiddleTier.DataStandard2/Base/MSSQL/0031-View-AcademicTimePeriodDim.sql:
```diff
@@ -0,0 +1,82 @@
+			CAST(Session.TermDescriptorId as VARCHAR),
+			CAST(GradingPeriod.GradingPeriodDescriptorId as VARCHAR),
+			CONVERT(VARCHAR, GradingPeriod.BeginDate, 112)
+		) as AcademicTimePeriodKey,
```
I recently learned that SQL Server 2016 introduced an updated version of [FORMATMESSAGE](https://docs.microsoft.com/en-us/sql/t-sql/functions/formatmessage-transact-sql?view=sql-server-ver15) allowing us to pass in any string for formatting. I think this is easier to read than the CONCAT statements that were used before, because you can see on line 16 exactly what the format is.

In PostgreSQL it is [FORMAT](https://www.postgresqltutorial.com/postgresql-format/) instead of `FORMATMESSAGE` and otherwise works identically.

stephenfuqua replied:
... and I just learned that there's an important difference between `CONCAT` and `FORMATMESSAGE`: `CONCAT` will automatically replace a `null` value with an empty string, but `FORMATMESSAGE` does not do so and will throw an error. Therefore `COALESCE()` or `ISNULL()` is necessary for nullable columns.

stephenfuqua commented on src/EdFi.AnalyticsMiddleTier.Tests/Dimensions/When_querying_the_SectionDim_view.cs:
```diff
@@ -71,7 +71,6 @@ public void Then_should_have_EducationalEnvironmentDescriptor()
             testResult.success.ShouldBe(true, testResult.errorMessage);
         }
 
-        [SuppressMessage("ReSharper", "InconsistentNaming")]
```
Redundant - see line 15 above.

stephenfuqua commented on src/EdFi.AnalyticsMiddleTier.DataStandard2/Base/MSSQL/0030-View-SectionDim-for-Student-Engagement.sql:
```diff
@@ -0,0 +1,69 @@
+﻿-- SPDX-License-Identifier: Apache-2.0
```
This screenshot shows the changes, when compared to the last alteration of `SectionDim`.

![image](https://user-images.githubusercontent.com/9324390/125685141-f19d1809-3e6e-44a2-ab42-361ef1991dfe.png)

The join to `CourseOffering` did not use all of the fields in the primary key. Adding the two additional fields allowed me to remove `DISTINCT` from the query.

On the first line, `SchoolId` is not nullable, therefore there is no need for `COALESCE`.

Note there are three new fields: `LocalEducationAgencyKey`, `CourseTitle`, and `SessionKey`.

Finally, there is a bug that I have not fixed, and will create a separate ticket for. `CourseTitle as SessionName` is not appropriate, as they are very different concepts. The correct column from the `Session` table was used in the DS31 and DS32 views.

stephenfuqua commented on src/EdFi.AnalyticsMiddleTier.DataStandard31/Base/MSSQL/0028-View-SectionDim-for-Student-Engagement.sql:
```diff
@@ -0,0 +1,108 @@
+END
+GO
+
+CREATE VIEW analytics.SectionDim AS
```
The comment on the DS2 version of this file applies equally to the changes you see here.

stephenfuqua commented on src/EdFi.AnalyticsMiddleTier.DataStandard32/Base/MSSQL/0028-View-SectionDim-for-Student-Engagement.sql:
```diff
@@ -0,0 +1,103 @@
+    DROP VIEW analytics.SectionDim
+END
+GO
+
```
The comment on the DS2 version of this file applies equally to the changes you see here.

DavidJGapCR commented:
If I compare the new SectionDim from DS 3.1 against the one from DS 3.2 with a tool like winmerge, it shows differences. Although it seems to be just formatting differences, so it's not important, these views should be, in most cases, identical. And if not, at least very similar.

DavidJGapCR commented:
In DS 2 project we have 2 scripts starting with 0030-<...>
And then on DS3.1 and 3.2 we have 2 scripts starting with 0028-<...>

stephenfuqua commented:
> If I compare the new SectionDim from DS 3.1 against the one from DS 3.2 with a tool like winmerge, it shows differences. Although it seems to be just formatting differences, so it's not important, these views should be, in most cases, identical. And if not, at least very similar.

I've updated DS32 to match DS31 with a fresh commit.

stephenfuqua commented:
> In DS 2 project we have 2 scripts starting with 0030-<...>
> And then on DS3.1 and 3.2 we have 2 scripts starting with 0028-<...>

These must have come in with a rebase. Although ultimately it doesn't matter, I'll rename these files to avoid numeric duplication. ... DONE.

jleiva-gap reviewed (changes requested):
View-AcademicTimePeriodDim.sql (DS 31 and DS 32). If I comment out the following lines some testcases should failed, but all is green.

- LastModifiedDate (46, 47) 
- Join - and (72,74,76,78)

View-StudentInternetAccessDim.sql (DS 31 and DS 32). If I comment out the following lines some testcases should failed, but all is green.
- Join - and (45, 56, 59,66, 72)

stephenfuqua commented:
> View-AcademicTimePeriodDim.sql (DS 31 and DS 32). If I comment out the following lines some testcases should failed, but all is green.
> 
> * LastModifiedDate (46, 47)
> * Join - and (72,74,76,78)
> 
> View-StudentInternetAccessDim.sql (DS 31 and DS 32). If I comment out the following lines some testcases should failed, but all is green.
> 
> * Join - and (45, 56, 59,66, 72)

DavidJGapCR commented:
I don't see StudentInternetAccessDim on DS 2 project. Is that Okay?

stephenfuqua commented:
> I don't see StudentInternetAccessDim on DS 2 project. Is that Okay?

Yes: the underlying table did not exist in Data Standard 2.

stephenfuqua commented:
@jleiva-gap @DavidJGapCR I've pushed additional commits to resolve the issues you have raised, please take a look when you have a few minutes.

jleiva-gap reviewed (changes requested):
View-AcademicTimePeriodDim.sql (DS 2). If I comment out the following lines some testcases should failed, but all is green.
- LastModifiedDate (46, 47, 48): To validate this case, I recommend combining the LastModifiedDate. For a record that is the SchoolYearType, in another that of Session and one more for GradingPeriod. In this way, when commenting on any of the lines, at least one of the test cases will fail. At this moment some columns have the same date, so even if a line is deleted, the result is the same.
Join - and (68)
- Join - and (72,74,76)

View-AcademicTimePeriodDim.sql (DS 31 and DS 32). If I comment out the following lines some testcases should failed, but all is green.
- LastModifiedDate (46, 47, 48): To validate this case, I recommend combining the LastModifiedDate. For a record that is the SchoolYearType, in another that of Session and one more for GradingPeriod. In this way, when commenting on any of the lines, at least one of the test cases will fail. At this moment some columns have the same date, so even if a line is deleted, the result is the same.
Join - and (64,68)
- Join - and (72,74,76,78)

stephenfuqua commented:
> View-AcademicTimePeriodDim.sql (DS 2). If I comment out the following lines some testcases should failed, but all is green.
> 
> * LastModifiedDate (46, 47, 48): To validate this case, I recommend combining the LastModifiedDate. For a record that is the SchoolYearType, in another that of Session and one more for GradingPeriod. In this way, when commenting on any of the lines, at least one of the test cases will fail. At this moment some columns have the same date, so even if a line is deleted, the result is the same.
>   Join - and (68)
> * Join - and (72,74,76)
> 
> View-AcademicTimePeriodDim.sql (DS 31 and DS 32). If I comment out the following lines some testcases should failed, but all is green.
> 
> * LastModifiedDate (46, 47, 48): To validate this case, I recommend combining the LastModifiedDate. For a record that is the SchoolYearType, in another that of Session and one more for GradingPeriod. In this way, when commenting on any of the lines, at least one of the test cases will fail. At this moment some columns have the same date, so even if a line is deleted, the result is the same.
>   Join - and (64,68)
> * Join - and (72,74,76,78)

I've corrected `LastModifiedDate`. The others are not practical - truthfully, I haven't found a way to get them to fail, although it should be possible. The problem is the intricate connection between all of the objects. Let's take the join for `GradingPeriod`, for example. If we try to make a test that fails when `SchoolId` is commented out, then we have to create an extra `GradingPeriod`. The foreign key for that extra grading period, when it connects to `SessionGradingPeriod`, has now just changed - so we have to create an extra `SessionGradingPeriod`. Oh, but the `SchoolId` is also part of the `Session` foreign key! So now we have to create an extra `Session`... and that just defeated the purpose of the test, because the earlier join on `Session` also fails because of the change to `SchoolId`.

Theoretically there is probably a way to get this to fail, but I'm not seeing it right now. So the next question - why include the join condition if we can't prove that it is needed? Two reasons: (1) the join should always include every element in the foreign key as a matter of principle, and (2) again, there's probably some intricate scenario that I'm failing to think of. Just by review alone, you can compare to the foreign keys and see that "* Join - and (72,74,76)" are all required.

```sql
ALTER TABLE [edfi].[SessionGradingPeriod]  WITH CHECK ADD  CONSTRAINT [FK_SessionGradingPeriod_GradingPeriod_SchoolId] FOREIGN KEY([GradingPeriodDescriptorId], [SchoolId], [BeginDate])
REFERENCES [edfi].[GradingPeriod] ([GradingPeriodDescriptorId], [SchoolId], [BeginDate])
```

jleiva-gap commented:
> 
> 
> > View-AcademicTimePeriodDim.sql (DS 2). If I comment out the following lines some testcases should failed, but all is green.
> > 
> > * LastModifiedDate (46, 47, 48): To validate this case, I recommend combining the LastModifiedDate. For a record that is the SchoolYearType, in another that of Session and one more for GradingPeriod. In this way, when commenting on any of the lines, at least one of the test cases will fail. At this moment some columns have the same date, so even if a line is deleted, the result is the same.
> >   Join - and (68)
> > * Join - and (72,74,76)
> > 
> > View-AcademicTimePeriodDim.sql (DS 31 and DS 32). If I comment out the following lines some testcases should failed, but all is green.
> > 
> > * LastModifiedDate (46, 47, 48): To validate this case, I recommend combining the LastModifiedDate. For a record that is the SchoolYearType, in another that of Session and one more for GradingPeriod. In this way, when commenting on any of the lines, at least one of the test cases will fail. At this moment some columns have the same date, so even if a line is deleted, the result is the same.
> >   Join - and (64,68)
> > * Join - and (72,74,76,78)
> 
> I've corrected `LastModifiedDate`. The others are not practical - truthfully, I haven't found a way to get them to fail, although it should be possible. The problem is the intricate connection between all of the objects. Let's take the join for `GradingPeriod`, for example. If we try to make a test that fails when `SchoolId` is commented out, then we have to create an extra `GradingPeriod`. The foreign key for that extra grading period, when it connects to `SessionGradingPeriod`, has now just changed - so we have to create an extra `SessionGradingPeriod`. Oh, but the `SchoolId` is also part of the `Session` foreign key! So now we have to create an extra `Session`... and that just defeated the purpose of the test, because the earlier join on `Session` also fails because of the change to `SchoolId`.
> 
> Theoretically there is probably a way to get this to fail, but I'm not seeing it right now. So the next question - why include the join condition if we can't prove that it is needed? Two reasons: (1) the join should always include every element in the foreign key as a matter of principle, and (2) again, there's probably some intricate scenario that I'm failing to think of. Just by review alone, you can compare to the foreign keys and see that "* Join - and (72,74,76)" are all required.
> 
> ```sql
> ALTER TABLE [edfi].[SessionGradingPeriod]  WITH CHECK ADD  CONSTRAINT [FK_SessionGradingPeriod_GradingPeriod_SchoolId] FOREIGN KEY([GradingPeriodDescriptorId], [SchoolId], [BeginDate])
> REFERENCES [edfi].[GradingPeriod] ([GradingPeriodDescriptorId], [SchoolId], [BeginDate])
> ```

In this case we check the join to verify that the columns are being included. In case a condition of the JOIN is deleted, the result of the query may vary. For example, if in this query I delete the SchoolYear, and there are records for the same school and SessionName with different SchoolYear, the data obtained may vary, even an account with a primary key could return more than one record. 

INNER JOIN
		edfi.SessionGradingPeriod
	ON
		Session.SchoolYear = SessionGradingPeriod.SchoolYear
	AND
		Session.SessionName = SessionGradingPeriod.SessionName
	AND
		Session.SchoolId = SessionGradingPeriod.SchoolId
		
But testing those cases requires having different combinations of column values. In these cases, in addition to the correct record, a record is inserted for each condition of the JOIN with a different value than expected. In this way, if the join is altered, it is validated that the expected condition is not met.

- edfi.SessionGradingPeriod: SchoolYear - SchoolId - Session name
- edfi.SessionGradingPeriod:  Different SchoolYear - SchoolId - Session name
- edfi.SessionGradingPeriod: SchoolYear - Different SchoolId - Session name
- edfi.SessionGradingPeriod: SchoolYear - SchoolId - Different Session name

stephenfuqua commented:
I think that the latest commits cover all outstanding cases.

jleiva-gap reviewed (approved):
The tests look good.
